sim/hunter/focus_fire.go

```diff
@@ -0,0 +1,100 @@
+	focusFireActionId := core.ActionID{SpellID: 428726}
+	focusFireFrenzyActionId := core.ActionID{SpellID: 428728}
+
+	// For tracking in timeline
+	hunterFrenzyAura := hunter.RegisterAura(core.Aura{
+		Label:     "Focus Fire Frenzy (Hunter)",
+		ActionID:  focusFireFrenzyActionId,
+		Duration:  time.Second * 10,
+		MaxStacks: 5,
+	})
+
+	hunterPetFrenzyAura := hunter.pet.RegisterAura(core.Aura{
+		Label:     "Focus Fire Frenzy",
+		ActionID:  focusFireFrenzyActionId,
+		Duration:  time.Second * 10,
+		MaxStacks: 5,
+		OnStacksChange: func(aura *core.Aura, sim *core.Simulation, oldStacks int32, newStacks int32) {
+			// revert previous attack speed bonus from stacks and apply new attack speed bonus
+			aura.Unit.MultiplyMeleeSpeed(sim, 1/(1+(0.06*float64(oldStacks))))
+			aura.Unit.MultiplyMeleeSpeed(sim, 1+(0.06*float64(newStacks)))
+
+  			// this is just so we can see the frenzy stacks working on the sim timeline
+			if !hunterFrenzyAura.IsActive() { hunterFrenzyAura.Activate(sim) }
+			hunterFrenzyAura.SetStacks(sim, newStacks)
+		},
+	})
+
+	hunterFocusFireAura := hunter.RegisterAura(core.Aura{
+		Label:    "Focus Fire",
+		ActionID: focusFireActionId,
+		Duration: time.Second * 20,
+		MaxStacks: 5,
+		OnStacksChange: func(aura *core.Aura, sim *core.Simulation, oldStacks, newStacks int32) {
+			aura.Unit.MultiplyMeleeSpeed(sim, 1/(1+(0.06*float64(oldStacks))))
+			aura.Unit.MultiplyMeleeSpeed(sim, 1+(0.06*float64(newStacks)))
+		},
+	})
+
+	hunter.pet.RegisterAura(core.Aura{
+		Label:    "Focus Fire Pet",
+		Duration: core.NeverExpires,
+		OnReset: func(aura *core.Aura, sim *core.Simulation) {
+			aura.Activate(sim)
+		},
+		OnSpellHitDealt: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+			if spell.ProcMask.Matches(core.ProcMaskMeleeMHAuto) {
+				if !hunterPetFrenzyAura.IsActive() {
+					hunterPetFrenzyAura.Activate(sim)
+				}
+
+				hunterPetFrenzyAura.AddStack(sim)
+				hunterPetFrenzyAura.Refresh(sim)
+				hunterFrenzyAura.Refresh(sim)
+			}
+		},
+	})
+
+	hunter.FocusFire = hunter.RegisterSpell(core.SpellConfig{
+		ActionID: focusFireActionId,
+
+		ManaCost: core.ManaCostOptions{
+			FlatCost: 0,						// TODO: Update when mana cost is known
+		},
+		Cast: core.CastConfig{
+			DefaultCast: core.Cast{
+				GCD: core.GCDDefault,
+			},
+			CD: core.Cooldown{
+				Timer:    hunter.NewTimer(),
+				Duration: time.Second * 15, 	// TODO: Update when cooldown is known (Currently using the cooldown value from Cataclysm)
+			},
+		},
+		ExtraCastCondition: func(sim *core.Simulation, target *core.Unit) bool {
+			return hunter.pet.IsEnabled() && (hunterPetFrenzyAura.GetStacks() > 0)
+		},
+		ApplyEffects: func(sim *core.Simulation, _ *core.Unit, _ *core.Spell) {
+			frenzyStacks := hunterPetFrenzyAura.GetStacks()
+			hunter.pet.AddFocus(sim, float64(4 * frenzyStacks), focusFireMetrics)
+
+			if !hunterFocusFireAura.IsActive() { hunterFocusFireAura.Activate(sim) }
```

Comment:
You should not check if its active here because by doing so you wont refresh its duration if it was already active

Reply:
Just calling activate will properly refresh the aura if it was already active

Reply:
I'll make this change, but in the future should I always call activate an aura before changing stacks if I want the buff to refresh or should I call aura.Refresh inside of the OnStackChange hook?

Reply:
Calling activate does not do anything to the stacks. It activates the aura if its not active or refreshes it if it is. Stack handling is done separately as auras have different rules for how their stacks work. In your case you want to always set stacks to the correct number after the activate